File: waitAll.py
import time
import asyncio
from threading import Thread
from multiprocessing import Process, Manager
from concurrent import futures
from operator import setitem

"""たらい回し関数"""
tarai = lambda x,y,z: y if x<=y else tarai(tarai(x-1,y,z),tarai(y-1,z,x),tarai(z-1,x,y))
"""フィボナッチ関数"""
fib = lambda n: n if n<2 else fib(n-1)+fib(n-2)
"""タイマーの経過時間"""
getTick = lambda t: time.time()-t

async def myFunc(taskName, i, v, timer):
	"""テスト関数"""
	ticks = []
	spans = []
	ticks.append(getTick(timer))
	await asyncio.sleep(v*1e-3*100)
	ticks.append(getTick(timer))
	spans.append(ticks[-1]-ticks[-2])
	tarai(12, 6, 0)
	ticks.append(getTick(timer))
	spans.append(ticks[-1]-ticks[-2])
	fibResult = fib(30+v)
	ticks.append(getTick(timer))
	spans.append(ticks[-1]-ticks[-2])
	return ["Python", taskName, str(i), str(v), str(fibResult)]+[f"{t:.3f}" for t in ticks+spans]
def myFuncSync(*args):
	"""テスト関数(同期)"""
	return asyncio.run(myFunc(*args))
def myFuncSyncRefs(refs, i, args):
	"""テスト関数(配列参照代入)"""
	refs[i] = myFuncSync(*args)

async def syncAll(taskName, values, timer, _):
	"""同期タスク"""
	return [await myFunc(taskName, i, v, timer) for i, v in enumerate(values)]

async def asyncAll(taskName, values, timer, _):
	"""非同期タスク(シングルスレッド)"""
	return await asyncio.gather(*(myFunc(taskName, i, v, timer) for i, v in enumerate(values)))

def procAll(Task):
	"""非同期タスク(低水準スレッド/プロセス)"""
	async def __taskAll(taskName, values, timer, _):
		result = [None]*len(values)
		if Task is Process: result = Manager().list(result)
		threads = [
			Task(target=myFuncSyncRefs, args=(result, i, (taskName, i, v, timer,),))
			for i, v in enumerate(values)
		]
		for t in threads: t.start()
		for t in threads: t.join()
		return list(result)
	return __taskAll

async def futureAll(taskName, values, timer, executor):
	"""非同期タスク(高水準スレッド/プロセス)"""
	return [f.result() for f in [executor.submit(myFuncSync, taskName, i, v, timer) for i, v in enumerate(values)]]
	# return list(executor.map(myFuncSync, values))

async def main():
	"""メイン処理"""
	result = [["#Lang", "Task", "Index", "Value", "Result", "Start", "Sleeped", "Taraied", "Fibed", "SleepTime", "TaraiTime", "FibTime"]]
	tasks = (
		("sync", syncAll, None),
		("async", asyncAll, None),
		("thread", procAll(Thread), None),
		("process", procAll(Process), None),
		("futureThread", futureAll, futures.ThreadPoolExecutor),
		("futureProcess", futureAll, futures.ProcessPoolExecutor),
	)
	values = list(range(10,0,-1))
	for taskName, task, Executor in tasks:
		if Executor is None: Executor = futures.ThreadPoolExecutor
		with Executor(max_workers=len(values)) as executor:
			timer = time.time()
			result.extend(await task(taskName, values, timer, executor))
			print(f"!{taskName}:{getTick(timer):.3f}")
	print("\n".join("\t".join(row) for row in result))

if __name__=="__main__":
	asyncio.run(main())

File: waitAll.r
library(parallel)

#' @title たらい回し関数
tarai<-function(x,y,z) ifelse(x<=y, y, tarai(tarai(x-1,y,z),tarai(y-1,z,x),tarai(z-1,x,y)))
#' @title フィボナッチ関数
fib<-function(n) ifelse(n<2, n, fib(n-1)+fib(n-2))
#' @title タイマーの経過時間
getTick<-function(t) (Sys.time()-t)

#' @title テスト関数
myFunc<-function(taskName, i, v, timer){
	ticks<-numeric(0)
	spans<-numeric(0)
	ticks<-c(ticks, getTick(timer))
	Sys.sleep(v*100*1e-3)
	ticks<-c(ticks, getTick(timer))
	spans<-c(spans, ticks[length(ticks)]-ticks[length(ticks)-1])
	tarai(12, 6, 0)
	ticks<-c(ticks, getTick(timer))
	spans<-c(spans, ticks[length(ticks)]-ticks[length(ticks)-1])
	fibResult<-0#fib(30+v)
	ticks<-c(ticks, getTick(timer))
	spans<-c(spans, ticks[length(ticks)]-ticks[length(ticks)-1])
	c("R", taskName, as.character(c(i, v , fibResult)), sprintf("%.3f", c(ticks, spans)))
}

#' @title 同期処理
syncAll<-function(taskName, values, timer){
	lapply(seq_along(values), function(i) myFunc(taskName, i, values[i], timer))
}

#' @title 非同期処理(パラレル(プロセス))
parallelAll<-function(taskName, values, timer) {
	cluster<-makeCluster(length(values))
	clusterExport(cluster, varlist = c("myFunc", "getTick", "tarai", "fib"))
	result<-parLapply(cluster, seq_along(values), function(i) myFunc(taskName, i, values[i], timer))
	stopCluster(cluster)
	result
}

#' @title メイン処理
main<-function(){
	result<-list(c("#Lang", "Task", "Index", "Value", "Result", "Start", "Sleeped", "Taraied", "Fibed", "SleepTime", "TaraiTime", "FibTime"))
	tasks<-list(
		list(Name="sync", Fn=syncAll),
		list(Name="parallel", Fn=parallelAll)
	)
	values<-10:1
	for(task in tasks){
		cat("!", task$Name, ": ", sep="")
		timer<-Sys.time()
		result<-c(result, task$Fn(task$Name, values, timer))
		cat(sprintf("%.3f\n", getTick(timer)))
	}
	cat(paste(sapply(result, function(row) paste(row, collapse="\t")), collapse="\n"), "\n")
}

main()
